xdsl/ir/core.py

```diff
@@ -63,9 +63,17 @@ class MLContext:
 
     allow_unregistered: bool = field(default=False)
 
-    _loaded_dialects: dict[str, Dialect] = field(init=False, default_factory=dict)
-    _loaded_ops: dict[str, type[Operation]] = field(init=False, default_factory=dict)
-    _loaded_attrs: dict[str, type[Attribute]] = field(init=False, default_factory=dict)
+    _loaded_dialects: dict[str, Dialect] = field(default_factory=dict)
+    _loaded_ops: dict[str, type[Operation]] = field(default_factory=dict)
+    _loaded_attrs: dict[str, type[Attribute]] = field(default_factory=dict)
+
+    def clone(self) -> MLContext:
+        return MLContext(
+            self.allow_unregistered,
+            self._loaded_dialects,
+            self._loaded_ops,
+            self._loaded_attrs,
```

Comment:
Isn't this a classic case of capturing? So, if I were to load more operations in the cloned context, I would also modify the original. I don't think we should implement it like this. I would prefer a deep clone in this first instance, and maybe later a shallow clone, if we feel this is needed.

Reply:
Good point, I think we can get away with just creating a new list for each of these.

Reply:
I guess a couple of `.copy()`s might do the trick?

Reply:
sure, not sure what the difference would be between `list(self._loaded_dialects)` and `self._loaded_dialects.copy()` would be in this case... I guess if the type changes from list to something else then the copy would still work but the list would fail, so copy is better.

Reply:
I agree with @webmiche. Maybe it's too early to see what is required by "core".
If it is something to consider for the interactive project, maybe tracking these TODOs/FIXMEs/etc in a gh project? Not sure what to suggest without impeding progress for @dshaaban01

Reply:
I would say that this is a useful thing to have independently of the interactive stuff, we should be able to copy an MLContext IMO